### DISCUSSION TITLE
Reduction function returning unique value

### DISCUSSION THREAD

**Discussion by zmoon** [pydata/xarray#7103] [Q&A] [answered]
Along a dimension or in a groupby group, I want to perform a reduction that checks that there is only one unique value in the group and returns it. I couldn't figure out how to apply the below with `xr.apply_ufunc()` (nor with groupby).

```python
import numpy as np
import xarray as xr

def the_unique(x):
    """Return _the_ non-NaN unique value or raise ValueError."""
    uns = np.unique(x)
    uns = uns[~np.isnan(uns)]
    if uns.size == 1:
        return uns[0]
    elif uns.size == 0:
        raise ValueError("no non-NaN values")
    else:
        raise ValueError("more than one unique value")

assert the_unique([1, 1, 1]) == 1

x = np.tile(np.arange(3)[:,np.newaxis], 3)  # each row has one unique value
np.testing.assert_equal(
    np.apply_along_axis(the_unique, 1, x),
    [0, 1, 2]
)

da = xr.DataArray(x, dims=["y", "x"])

# xr.apply_ufunc(...) ?
```

Example situation:
- going pandas -> xarray with table of time series at different points, in order to reduce duplication for variables that (should) only vary in space not time

**Comment by dcherian** [comment, marked as answer]
This should work with `reduce`? Also `pd.unique` is usually faster than `np.unique` and won't return multiple NaNs

For `apply_ufunc`:
1. `input_core_dims` should be the dimension you're reducing
2. I think you have no `output_core_dims`
3. To loop over an axis like `apply_along_axis` use `vectorize=True`.
You might even try
```
def the_unique(x):
	x = x[~np.isnan(x)]
	uniq = x[0]
	assert (x == uniq).all()
	return x
```

**Reply by zmoon** [reply]
Thanks @dcherian, I was able to get it to work with `apply_ufunc` by adding `vectorize=True` (I had `input_code_dims` as the dimension to reduce already). And I was able to use with groupby reduce by just adding `axis=None` to the signature (+ using the `pd` functions).

And thanks for the `pd.unique` tip, using that and `pd.isnull` allows this to work with non-numeric.

I will compare perf a bit, the code you posted looks faster, and maybe easier to make into a vectorized version that would use `axis`.
